File: answers/lintcode/Longest Common Subsequence.cpp
//@level Medium 
//@type LintCode Copyright Longest Common Subsequence Dynamic Programming
//@result Accepted Total Runtime: 10 ms 100% test cases passed.

class Solution {
public:
    /**
     * @param A, B: Two strings.
     * @return: The length of longest common subsequence of A and B.
     */
    int longestCommonSubsequence(string A, string B) {
        // write your code here
		vector<vector<int> > dp(A.length(), vector<int> (B.length(), 0));
		for (size_t i = 0; i < A.length(); ++ i) {
			for (size_t j = 0; j < B.length(); ++ j) {
				if (A[i] == B[j]) {
					dp[i][j] = (0 == i || 0 == j) ? 1 : dp[i - 1][j - 1] + 1;
				}
				else if (i > 0 && j > 0) {
					dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
				}
			}
		}
		return dp.size() ? dp[dp.size() - 1][dp[dp.size() - 1].size() - 1] : 0;
    }
};

File: answers/leetcode/Longest Common Subsequence/Longest Common Subsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int[][] dp = new int[text1.length()][text2.length()];
        for (int i = 0; i < text1.length(); i++) {
            for (int j = 0; j < text2.length(); j++) {
                if (i > 0) {
                    dp[i][j] = Math.max(dp[i][j], dp[i - 1][j]);
                }
                if (j > 0) {
                    dp[i][j] = Math.max(dp[i][j], dp[i][j - 1]);
                }
                if (text1.charAt(i) == text2.charAt(j)) {
                    dp[i][j] = Math.max(dp[i][j], (i > 0 && j > 0 ? dp[i - 1][j - 1] : 0) + 1);
                }
            }
        }
        return dp[text1.length() - 1][text2.length() - 1];
    }
}